### ISSUE TITLE
`gatsby-plugin-sharp` / `gatsby-remark-images`: Set destination directory with destinationDir option

### ISSUE THREAD

**Issue by karlhorky** [gatsbyjs/gatsby#27153]
## Summary

For the plugins `gatsby-plugin-sharp` and/or `gatsby-remark-images` (whichever makes sense and is necessary), an option `destinationDir` (string or function) that works similarly to [the option in `gatsby-remark-copy-linked-files`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-copy-linked-files#advanced-usage)...

```js
// save `my-awesome-pdf.pdf` to `public/downloads/2a0039f3a61f4510f41678438e4c863a/my-awesome-pdf.pdf`
destinationDir: f => `downloads/${f.hash}/${f.name}`

// save `my-awesome-pdf.pdf` to `public/downloads/2a0039f3a61f4510f41678438e4c863a-my-awesome-pdf.pdf`
destinationDir: f => `downloads/${f.hash}-${f.name}`

// save `my-awesome-pdf.pdf` to `public/my-awesome-pdf/2a0039f3a61f4510f41678438e4c863a.pdf`
destinationDir: f => `${f.name}/${f.hash}`
```

...also with the new `relativeDirectory` property in the returned file object:

```js
// save `my-awesome-pdf.pdf` to `public/original-path-in-src-pages/my-awesome-pdf.pdf`
destinationDir: f => `${f.relativeDirectory}/${f.name}`,
```

### Basic example

See examples above.

### Motivation

The structure of the files in the output `public` folder is more configurable and using the `relativeDirectory` option, can also match 1 to 1 the folder structure of the input `src/pages` directory. This makes the output easier to reason about and also easier to build tooling upon.

**Comment by sidharthachatterjee** [comment]
This makes a lot of sense. Happy to receive a PR adding this!

**Comment by karlhorky** [comment]
Nice! Could you weigh in on where the changes should be made? In `gatsby-plugin-sharp`? Or `gatsby-remark-images` / `gatsby-images`? Or all three?

**Comment by wardpeet** [comment]
@karlhorky could you tell us more about tooling that would be beneficial here? I feel that adding an option `destinationDir` might cause more problems than it fixes.

**Comment by karlhorky** [comment]
@wardpeet in lieu of an Asset Manifest (see https://github.com/gatsbyjs/gatsby/issues/20745), I would like the built assets to better match the inputs (in order to achieve server-side authentication and access control - eg. https://github.com/gatsbyjs/gatsby/issues/1100):

```
input:
src/pages/original/path/my-awesome-image.jpg

output:
public/original/path/my-awesome-image-83j4g898hg98sheijf0239ijf9djsf9h39h98hre.jpg
```

> will have issues with caching as we tell people to make these files immutable

This is a good point, and maybe also something that should be applied to [the `destinationDir` callback in `gatsby-remark-copy-linked-files`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-copy-linked-files#advanced-usage).

Maybe one way of doing this would be to require and validate that the content hash be used somewhere in the filename? We could do this by running the function and making sure that the argument is being used, à la:

https://github.com/gatsbyjs/gatsby/blob/6270c3d09ba5abfcd73197a3632bdbc737e96768/packages/gatsby-remark-copy-linked-files/src/index.js#L37

**Comment by karlhorky** [comment]
not stale

**Comment by karlhorky** [comment]
not stale

**Comment by karlhorky** [comment]
lol such a stale bot

not stale

**Comment by wardpeet** [comment]
> But instead of validating that EITHER the name OR the hash is used, the validation could be stricter - that the hash is used.

That's interesting, I'm still hesitant to enable this behavior as I don't fully understand the complications that this will bring. Technically we could start with adding outputDir to `fluid`, `fixed`, ...

Would it help if you got a list of assets per route?

**Comment by karlhorky** [comment]
> That's interesting, I'm still hesitant to enable this behavior as I don't fully understand the complications that this will bring.

Well, as long as the `hash` (and also `versionHash`, I'm just realizing) is included in the name somehow, I don't think this would cause issues. It would need to generate multiple different:

**Original Path**

```
pages/git-github/002-dvcs.png
```

**Before**

```
static/289ee13ff077319387caa5cb396c6cea/2a79a/002-dvcs.png
static/289ee13ff077319387caa5cb396c6cea/3b59c/002-dvcs.png
static/289ee13ff077319387caa5cb396c6cea/74866/002-dvcs.png
```

**After (without `relativeDirectory`)**

Here the first `static` part of the URL stays the same (cannot be changed through this API).

The name `destinationDir` doesn't actually make sense here, because it is actually ONLY changing the filename. But `destinationDir` also doesn't make sense in `gatsby-remark-copy-linked-files`...

Config: `` destinationDir: f => `${f.name}-${f.hash}-${f.versionHash}` ``

```
static/289ee13ff077319387caa5cb396c6cea/2a79a/002-dvcs-289ee13ff077319387caa5cb396c6cea-2a79a.png
static/289ee13ff077319387caa5cb396c6cea/3b59c/002-dvcs-289ee13ff077319387caa5cb396c6cea-3b59c.png
static/289ee13ff077319387caa5cb396c6cea/74866/002-dvcs-289ee13ff077319387caa5cb396c6cea-74866.png
```

**After (with `relativeDirectory`)**

Config: `` destinationDir: f => `${f.relativeDirectory}/${f.name}/${f.versionHash}/${f.name}-${f.hash}` ``

```
public/git-github/002-dvcs.png/2a79a/002-dvcs-289ee13ff077319387caa5cb396c6cea.png
public/git-github/002-dvcs.png/3b59c/002-dvcs-289ee13ff077319387caa5cb396c6cea.png
public/git-github/002-dvcs.png/74866/002-dvcs-289ee13ff077319387caa5cb396c6cea.png
```

**Comment by karlhorky** [comment]
> Would it help if you got a list of assets per route?

Sure, that would solve it a different way. That would be my request over at https://github.com/gatsbyjs/gatsby/issues/20745

**Comment by wardpeet** [comment]
I feel there are more use-cases for a manifest file than there might be for renaming files. This would indicate that these files are part of a route.

**Comment by karlhorky** [comment]
Yeah definitely, I think there are some use cases for sure for a full asset manifest.

Right now [I'm using the information in `webpack.stats.json`](https://github.com/gatsbyjs/gatsby/issues/20745#issuecomment-577702094), which gets me some things. But a unified file that contained everything in the webpack stats and also these new files would be great.

Back then I made [a list](https://github.com/gatsbyjs/gatsby/issues/20745#issuecomment-577709357) of all the things that I found missing from `webpack.stats.json`.
